core/src/main/scala/kafka/server/DynamicBrokerConfig.scala

```diff
@@ -358,3 +364,52 @@ class DynamicBrokerConfig(private val kafkaConfig: KafkaConfig) extends Logging
     }
   }
 }
+
+object DynamicLogConfig {
+  // Exclude message.format.version for now since we need to check that the version
+  // is supported on all brokers in the cluster.
+  val ExcludedConfigs = Set(KafkaConfig.LogMessageFormatVersionProp)
+
+  val ReconfigurableConfigs = LogConfig.TopicConfigSynonyms.values.toSet -- ExcludedConfigs
+  val KafkaConfigToLogConfigName = LogConfig.TopicConfigSynonyms.map { case (k, v) => (v, k) }
+}
+class DynamicLogConfig(logManager: LogManager) extends Reconfigurable with Logging {
+
+  override def configure(configs: util.Map[String, _]): Unit = {}
+
+  override def reconfigurableConfigs(): util.Set[String] = {
+    DynamicLogConfig.ReconfigurableConfigs.asJava
+  }
+
+  override def validateReconfiguration(configs: util.Map[String, _]): Boolean = {
+    // For update of topic config overrides, only config names and types are validated
+    // Names and types have already been validated. For consistency with topic config
+    // validation, no additional validation is performed.
+    true
+  }
+
+  override def reconfigure(configs: util.Map[String, _]): Unit = {
+    val currentLogConfig = logManager.currentDefaultConfig
+    val newBrokerDefaults = new util.HashMap[String, Object](currentLogConfig.originals)
+    configs.asScala.filterKeys(DynamicLogConfig.ReconfigurableConfigs.contains).foreach { case (k, v) =>
+      if (v != null) {
+        DynamicLogConfig.KafkaConfigToLogConfigName.get(k).foreach { configName =>
+          newBrokerDefaults.put(configName, v.asInstanceOf[AnyRef])
+        }
+      }
+    }
+
+    logManager.reconfigureDefaultLogConfig(LogConfig(newBrokerDefaults))
+
+    logManager.allLogs.foreach { log =>
+      val props = new Properties
+      newBrokerDefaults.asScala.foreach { case (k, v) => props.put(k, v) }  // add new broker defaults
+      log.config.originals.asScala.filterKeys(log.config.overriddenConfigs.contains).foreach { case (k, v) =>
+        props.put(k, v) // add topic overrides
+      }
+
+      val logConfig = LogConfig(props)
```

Comment:
Doesn't seem like we actually need a `Properties` object. I was wondering if using a `Map` would be more concise (e.g. we'd be able to use `putAll`). For example:

```scala
      val props = mutable.Map.empty[Any, Any]
      props ++= newBrokerDefaults.asScala
      props ++= log.config.originals.asScala.filterKeys(log.config.overriddenConfigs.contains)
      val logConfig = LogConfig(props.asJava)
```